mlir-sycl/lib/Transforms/Inliner.cpp

```diff
@@ -701,6 +701,14 @@ bool Inliner::inlineCallsInSCC(Inliner &Inliner, CGUseList &UseList,
     DidSomething = true;
     ++NumInlinedCalls;
 
+    // Move all AllocaOp to the beginning of the caller function.
```

Comment:
By generating `sycl.addrspacecast()` instead of `polygeist.pointer2memref(llvm.addrspacecast(polygeist.memref2pointer()))`, there are more callsites get inlined. 
`Basic/linear-sub_group.cpp` failed with an extra callsite inlined. The failure can be worked around by disabling runtime optimizations (`SYCL_PROGRAM_COMPILE_OPTIONS="-cl-opt-disable"`). 
Or `Basic/linear-sub_group.cpp` can pass when all alloca operations are at the beginning of a function. Seems like there are runtime optimizations that rely on all alloca operations at the beginning of a function.

Reply:
Was worth a try. Thanks for trying it.